documentation/Configs.md
### Configs

#### Settings Usage

you can programmatically set it in your code. Recommended you just change it in the settings or env config 

``` javascript
self.settings = require('./configs/settings.js').get()
self.settings = require('./configs/settings.js').set({
  'swagger':false
})
```

#### Seo.js

Example below of how totally control the seo of a dynamic route. In the hook if you chose to create one you have access to the `self` obj for the entire system, `data` object which will get sent to your templates & lastly if you need to call the callback to pass a error first or the data second. if there is a error the system will revert to the default seo. Note the data object will hold the values of the route before you do anything to it for example anything on the query or the params. In the env configs you can deem how you want to render by which template engine. Default - `lodash` which can handles es6 and ejs-like templating. If you wanted to you could switch right from the start to ejs and the templates are already compatible 

``` javascript
'/blog/view/:id': {
    title: '<%=  blog.title  %> -  <%=  blog.user.profile.name %> ',
    keywords: '<%=  blog.tags  %>',
    description: '<%=  blog.content  %> ',
    ogUrl: '<%=  path  %>',
    twitterUrl: '<%=  path  %>',
    canonical: '<%=  path  %>',
    ogTitle: '<%=  blog.title  %> -  <%=  blog.user.profile.name  %> ',
    twitterTitle: '<%=  blog.title  %> -  <%=  blog.user.profile.name  %> ',
    ogDescription: '<%=  blog.content  %> ',
    twitterDescription: '<%=  blog.content  %> ',
    hook: function (self, data, cb) {
      data.blog = {
        tags: ['Add', 'Tags', 'To Blog', 'Mean Stack JS']
      }
      self.models.blog.findOne({_id: data.params.id}).populate('user').then(function (blog) {
        data.blog = _.assign(data.blog, blog)
        cb(null, data)
      }).catch(function (err) {
        cb(err)
      })
    }
  }
```
#### Environment.js

``` javascript
self.environment = require('./server/environment.js').get()
self.environment = require('./server/environment.js').set('development')
// or by CMD
export NODE_ENV=test // LINUX
set NODE_ENE=test // WINDOWS
```

#### Configs are based off env

``` javascript
// Development example
var mongodbUri = process.env.DB_PORT_27017_TCP_ADDR || process.env.MONGODB || process.env.MONGOLAB_URI || 'mongodb://localhost/dev'
module.exports = {
  minify: 'default', // 'concat' all files or 'minify' concat and minfy  or 'default' leave as is
  html: {
    title: 'Development MEANSTACKJS'
  },
  logger: 'dev',
  cdn: process.env.CDN || false,
  buildreq: {
    console: true
  },
  maxcdn: {
    zoneId: process.env.MAXCDN_ZONE_ID || false
  },
  mongoexpress: {
    port: process.env.MONGOEXPRESSPORT || 8081
  },
  socketio: {
    port: process.env.SOCKETIOPORT || 8282
  },
  http: {
    active: true,
    port: process.env.PORT || 3000
  },
  https: {
    active: false,
    port: process.env.HTTPSPORT || 3043,
    key: './configs/certificates/keyExample.pem',
    cert: './configs/certificates/certExample.pem'
  },
  throttle: {
    rateLimit: {
      ttl: 600,
      max: 20000
    },
    mongoose: {
      uri: mongodbUri
    }
  },
  mongodb: {
    uri: mongodbUri,
    // Database options that will be passed directly to mongoose.connect
    // Below are some examples.
    // See http://mongodb.github.io/node-mongodb-native/driver-articles/mongoclient.html#mongoclient-connect-options
    // and http://mongoosejs.com/docs/connections.html for more information

    options: {
      // server: {
      //   socketOptions: {
      //     keepAlive: 1
      //   },
      //   poolSize: 5
      // },
      // replset: {
      //   rs_name: 'myReplicaSet',
      //   poolSize: 5
      // },
      db: {
        w: 1,
        numberOfRetries: 2
      }
    }
  },
  agendash: {
    active: true,
    options: {
      db: {
        address: mongodbUri
      }
    }
  }
}

```

#### Settings.js

We now have included [Dotenv](https://www.npmjs.com/package/dotenv) as you see below. In doing this it purposely fails silently because we want you to never source that info in github for security reasons. Please keep your configs safe and out of the public eye. `minify` is now configurable and no longer based on our env names so you have total control of your env names and how they build.

Note: assets are in reverse order of which there loaded

``` javascript
require('dotenv').config({silent: true})

var path = require('path')
var _ = require('lodash')
var environment = require('./environment.js').get()
var baseLine = {
  app: {
    name: 'MeanStackJS'
  },
  minify: 'default',
  render: {
    cli: 'lodash', // __ or ejs or lodash.
    seo: 'lodash', // ejs or lodash. default is lodash
    lodash: {
      options: {} // https://lodash.com/docs#template
    },
    ejs: {
      options: {} // https://www.npmjs.com/package/ejs#options
    }
  },
  env: environment,
  // Root path of server
  root: path.join(__dirname, '/../../..'),
  // Server IP
  ip: process.env.IP || '0.0.0.0',
  hostname: process.env.HOST || process.env.HOSTNAME || 'localhost',
  // Enable Swagger.io at localhost:[port]/api/
  swagger: true,
  // Enable the use of babel for ES6
  babel: {
    options: {
      // NOTE you must install the proper package to use here
      presets: [
        'babel-preset-es2015'
      ],
      plugins: [
        'transform-class-properties'
      ]
    },
    folder: 'dist',
    active: false
  },
  // Plato
  plato: {
    title: 'mean stack',
    eslint: {
      lastsemic: true,
      asi: true
    }
  },
  // JWT Object https://github.com/auth0/node-jsonwebtoken
  jwt: {
    // is used to compute a JWT SIGN
    secret: 'MEANSTACKJS',
    options: {
      expiresIn: 60 * 120 // 60 seconds * 120  = 2 hours
    }
  },
  // is used to compute a session hash
  sessionSecret: 'MEANSTACKJS',
  // The name of the MongoDB collection to store sessions in
  sessionCollection: 'sessions',
  // The session cookie settings
  sessionCookie: {
    path: '/',
    httpOnly: true,
    // If secure is set to true then it will cause the cookie to be set
    // only when SSL-enabled (HTTPS) is used, and otherwise it won't
    // set a cookie. 'true' is recommended yet it requires the above
    // mentioned pre-requisite.
    secure: false,
    // Only set the maxAge to null if the cookie shouldn't be expired
    // at all. The cookie will expunge when the browser is closed.
    maxAge: null
  },
  sessionName: 'session.id',
  // Supports MAX CDN
  maxcdn: {
    companyAlias: process.env.MAXCDN_COMPANY_ALIAS || '',
    consumerKey: process.env.MAXCDN_CONSUMER_KEY || '',
    consumerSecret: process.env.MAXCDN_CONSUMER_SECRET || ''
  },
  // SEO - Default html setup
  googleAnalytics: 'UA-71654331-1',
  html: {
    title: 'Mean Stack JS Demo',
    keywords: 'MEAN, MEANSTACKJS, mongodb, expressjs, angularjs,nodejs, javascript',
    description: 'Mean Stack JS was built for ease of use with javascript at its core. MeanStackJS is a full stack javascript framework that will give you the power to develop web applications',
    ogUrl: 'https://meanstackjs.herokuapp.com/',
    ogType: 'website',
    ogTitle: 'Mean Stack JS Demo',
    ogDescription: 'Mean Stack JS was built for ease of use with javascript at its core. MeanStackJS is a full stack javascript framework that will give you the power to develop web applications',
    ogImage: 'http://meanstackjs.com/images/logo/header.png',
    fbAppId: '1610630462580116',
    twitterCreator: '@greenpioneerdev',
    twitterCard: 'summary_large_image',
    twitterTitle: 'Mean Stack JS Demo',
    twitterDescription: 'Mean Stack JS was built for ease of use with javascript at its core. MeanStackJS is a full stack javascript framework that will give you the power to develop web applications',
    twitterUrl: 'https://meanstackjs.herokuapp.com/',
    twitterImage: 'http://meanstackjs.com/images/logo/header.png',
    twitterSite: '@meanstackjs',
    canonical: 'https://meanstackjs.herokuapp.com/',
    author: 'Green Pioneer Solutions'
  },
  seo: require('./seo.js'),
  // AGGREGATION
  // components -  Needs to be manually added below
  // modules - aggregated automatically
  // images - manually called in files
  // styles - manually called  & automatically compiles the global style scss in COMPILED Folder
  // uploads - Automatic uploads to be manually called in the files
  // USE EXTERNAL FILES - 'https://ajax.googleapis.com/ajax/libs/jquery/1.11.3/jquery.min.js',
  // OR USE INTERNAL FILES - '/bower_components/jquery/dist/jquery.js'
  assets: {
    js: [
      '/angular-jwt/dist/angular-jwt.js',
      '/socket.io-client/socket.io.js',
      '/ng-file-upload/dist/ng-file-upload-all.js',
      '/angular-mocks/angular-mocks.js',
      '/angular-cookies/angular-cookies.js',
      '/angular-sanitize/angular-sanitize.js',
      '/angular-animate/angular-animate.js',
      '/angular-resource/angular-resource.js',
      '/angular-ui-router/release/angular-ui-router.js',
      '/angular-ui-bootstrap/dist/ui-bootstrap-tpls.js',
      '/angular-moment/angular-moment.js',
      '/moment/moment.js',
      '/lodash/lodash.js',
      '/toastr/toastr.js',
      '/bootstrap-sass/assets/javascripts/bootstrap.js',
      '/angular/angular.js',
      '/jquery/dist/jquery.js'
    ],
    css: [
      '/styles/compiled/global.style.css',
      '/toastr/build/toastr.css'
    ],
    copy: [
      '/font-awesome/fonts/fontawesome-webfont.woff2',
      '/font-awesome/fonts/fontawesome-webfont.ttf'
    ]
  },
  bodyparser: {
    json: {limit: '100kb'},
    urlencoded: {limit: '100kb', extended: true}
  },
  helmet: {
    // https://github.com/helmetjs/helmet
  },
  expresValidator: {
    customValidators: {
      isArray: function (value) { // req.assert('param', 'Invalid Param').isArray()
        return _.isObject(value)
      },
      isObject: function (value) { // req.assert('param', 'Invalid Param').isObject()
        return _.isObject(value)
      },
      isString: function (value) { // req.assert('param', 'Invalid Param').isString()
        return _.isString(value)
      },
      isRegExp: function (value) { // req.assert('param', 'Invalid Param').isRegExp()
        return _.isRegExp(value)
      },
      isEmpty: function (value) { // req.assert('param', 'Invalid Param').isEmpty()
        return _.isEmpty(value)
      },
      gte: function (param, num) { // req.assert('param', 'Invalid Param').gte(5)
        return _.gte(param, num)
      },
      lte: function (param, num) { // req.assert('param', 'Invalid Param').lte(5)
        return _.lte(param, num)
      },
      gt: function (param, num) { // req.assert('param', 'Invalid Param').gt(5)
        return _.gt(param, num)
      },
      lt: function (param, num) { // req.assert('param', 'Invalid Param').lt(5)
        return _.lt(param, num)
      }
    },
    customSanitizers: {
      toArray: function (value) { // req.sanitize('postparam').toArray()
        return _.toArray(value)
      },
      toFinite: function (value) { // req.sanitize('postparam').toFinite()
        return _.toFinite(value)
      },
      toLength: function (value) { // req.sanitize('postparam').toLength()
        return _.toLength(value)
      },
      toPlainObject: function (value) { // req.sanitize('postparam').toPlainObject()
        return _.toPlainObject(value)
      },
      toString: function (value) { // req.sanitize('postparam').toString()
        return _.toString(value)
      }
    },
    errorFormatter: function (param, msg, value) {
      var namespace = param.split('.')
      var root = namespace.shift()
      var formParam = root

      while (namespace.length) {
        formParam += '[' + namespace.shift() + ']'
      }
      return {
        param: formParam,
        msg: msg,
        value: value
      }
    }
  },
  buildreq: {
    console: true,
    response: {
      method: 'get',
      data: {},
      user: {},
      count: 0,
      hostname: '',
      type: '',
      actions: {
        prev: false,
        next: false,
        reload: false
      },
      delete: ['error']
    },
    query: {
      sort: '',
      limit: 20,
      select: '',
      filter: {},
      populateId: 'user',
      populateItems: '',
      lean: false,
      skip: 0,
      where: '',
      gt: 1,
      lt: 0,
      in: [],
      equal: '',
      errorMessage: 'Unknown Value'
    },
    routing: {
      schema: true,
      url: '/api/v1/',
      build: true
    }
  },
  email: {
    templates: {
      welcome: {
        subject: 'Welcome to Mean Stack JS',
        text: function (username) {
          return 'Hi ' + username + ',\n\n' +
          'Thanks for signing up for Mean Stack JS.\n\n' +
          'If you have any questions about the site, you can reply to this ' +
          'email.\n\n' +
          '— Mean Stack JS'
        }
      },
      reset: {
        subject: 'Reset your password on MEANSTACKJS ',
        text: function (email) {
          return 'Hello,\n\n' +
          'This is a confirmation that the password for your account ' + email + ' has just been changed.\n'
        }
      },
      forgot: {
        subject: 'Welcome to Mean Stack JS',
        text: function (host, token) {
          return 'You are receiving this email because you (or someone else) have requested the reset of the password for your account.\n\n' +
          'Please click on the following link, or paste this into your browser to complete the process:\n\n' +
          'http://' + host + '/reset/' + token + '\n\n' +
          'If you did not request this, please ignore this email and your password will remain unchanged.\n'
        }
      }
    },
    from: 'MEANSTACKJS@localhost.com',
    error: 'MEANSTACKJS@localhost.com',
    connect: {
      host: 'smtp.mandrillapp.com', // Gmail, SMTP
      port: '587',
      auth: {
        user: 'hackathonstarterdemo',
        pass: 'E1K950_ydLR4mHw12a0ldA'
      }
    }
  }
}
if (environment === 'test') {
  baseLine = _.assign(baseLine, require('./environments/test.js'))
} else if (environment === 'production') {
  baseLine = _.assign(baseLine, require('./environments/production.js'))
} else if (environment === 'nightwatch') {
  baseLine = _.assign(baseLine, require('./environments/nightwatch.js'))
} else {
  baseLine = _.assign(baseLine, require('./environments/development.js'))
}

exports.get = function (env) {
  return baseLine
}
exports.set = function (identifer, value) {
  baseLine[identifer] = value
  return baseLine
}

```




documentation/Changelog.md
### Changelog

#### Mar 12th 18 - v1.9.0
#294 - Creating Scripts To Automate Documentation
#293 - File to import not found or unreadable: bootstrap
#292 - Update Dev Dependencies
#285 - Livereload Support HTTPS
#284 - Socket Support HTTP & HTTPS
#126 - Move Frontend Dep to NPM

#### Mar 6th 18 - v1.8.2
#291 Update Dependencies 

#### Feb 18th 18 - v1.8.1
#289 Fix meanstackjs loading problem due to too old node-sass package @hillkim7 
#288 Enhance configurable environment variables @shawncampbell 

#### Aug 1st 17 - v1.8.0
#280 Add settings for cache & security
#278 For validation errors after password reset
#277 Expose Socket.io to every Module enhancement
#276 Turn Http & Https on at same time enhancement
#275 Close mongoose connections on shutdown enhancement
#274 Update the proxy enhancement
#259 Add Local Storage Factory enhancement
#258 Support Yarn enhancement
#245 Update Logout

#### Feb 19th 17 - v1.7.0
#220 Add other Auths enhancement
#219 Update to GPS style enhancement
#218 abstract out routes & tools
#217 Add Better Logging enhancement
#211 Load Mongoose Models Sooner enhancement Ideas Investigate
#207 Replace Names bug
#198 Config Error Logger enhancement


#### Feb 6th 17 - v1.6.0
#210 Update users urls
#209 Add API Key Support enhancement
#208 can't handle production external URLs bug Submitted by @oshirodk
#206 Set all Deps to set versions bug
#204 Update Error Logger enhancement
#200 Add system status api enhancement
#199  Debug Front Info Page enhancement
#192 Link Account enhancement
#190 Add Debug Page


#### Jan 2nd 17 - v1.5.0
# Release 1.5.0

[b8243d3125072726b184e96cd2d01e0088466d8f](https://github.com/greenpioneersolutions/meanstackjs/commit/b8243d3125072726b184e96cd2d01e0088466d8f)
- Error Handling
- CLI upgrade
- New Scripts
- Error Dashboard
- Dotenv Integrated
- Custom Environments
- Minify Config

#181 
#179
#178
#177
#175
#174
#173 
#172
#171
#170
#169
#167
#165
#164
#161
#159 
#158
#157
#156 
#153


#### Dec 23rd 16 - v1.4.1
#161 


#### Dec 23rd 16 - v1.4.0
#148
#147
#146
#144
#143
#123


#### Nov 20th 16 - v1.3.0
#142
#141
#140 
#139 
#131 


#### Nov 5th 16 - v1.2.0
#127
#128
#129
#130
#132
#133
#134
#135
#136
#137


#### Oct 17th 16 - v1.1.0
#105
#108 
#109
#112 
#114
#115
#116
#117
#118
#119
#120
#121
#122
#123
#124
#125


#### Oct 17th 16 - v1.0.1
#113 Pushed the patch for bower resolution


#### Sep 12th 16 - v1.0.0
It is now officially ready for use 1.0.0 !!!!

More Documentation to come here later


#### Sep 6th 16 - v0.5.0
#96 Updated Stack , Testing , Cli & Comments
#95 Switched from Swig to EJS
#94 Resolve to $state.go 
#93 IsAuthorized Middleware
#92 Remove routing Buildreq

NOTE- VERSION 1.0 will be published within the week


#### Aug 17th 16 - v0.4.3
#90 Add MaxCDN enhancement
#88 Fix Iphone Error bug
#87 Scss Support ENV enhancement
#86 Add The Ability to Config Domain Change enhancement
#85 Fix: Redirect on Errors bug
#84 Add: Last Login to User Schema enhancement
#82 Update: the CLI Templates bug
#81 Contrib: Add Timestamp enhancement help wanted
#80 Add Code Wake
#79 Contrib: Add Full Admin Dashboard enhancement help wanted


#### Jul 26th 16 - v0.4.2
#5 Documentation & updates
#77  Add Simple Admin Portal & tests
#76 Fix mongoose connect error
#78 Fix all exports /modules
#75 style update


#### Jul 26th 16 - V0.4.1
#72 Add Https
#74 Add Mean Seo
#7 Add Security


#### Jul 26th 16 - v0.4.0
#66 Fix Agenda
#67 update seo user & system
#68 Add custom express validators
#69 Add Error Handler
#70 Add Backup & Restore job for mongo
#71 Add Site map
#44 Update bower


#### May 5th 16 - v0.3.3
#63 Update license
#64 Update Contributing
#59 Update the commands cli
#58 move the user factory
#57 update  frontend test cases
#56 add agenda 
#54 Fixed the models
#51 add more e2e testing
#48 Update npm packages
#43 Fixed cli


#### Apr 27th 16 - v0.3.2
#53 & #38  update index page
#54  fix the models , now it allows us to have multiple models
#48 update the npm list 
#51  add a login tests & general test
#41 add simple ng-enter directive
#52 add gitter badge
#49  & #36 update user module


#### Apr 20th 16 - v0.3.1
#35 added frontend test cases
#43 fixed cli error
#32 added travis.ci


#### Apr 19th 16 - v0.3.0
#1 added nightwatch.js e2e
#24 added mongo express server to give you admin db view in dev mode
#26 create and update the environment & settings js
#27 added plato to analyze code at localhost:3000/plato
#28  can check the reporter of night watch at localhost:3000/e2e
#30 Updated & Reorder the server and client side code


#### Apr 15th 16 - v0.2.1
#20 Added Babel for es6 support on the server side
#22 Added swagger at localhost:3000/api/
#23 Added Dockerfile


#### Apr 12th 16 - v0.2.0
## #19 did a major Refactor

Created Run Module - to run all type of servers ( socketio , mean  , livereload ..etc)
Refactor Register.js 
Refactor Mean.Sever.js 

Updated Commands
Updated Models
Updated Routes
Updated Controllers


#### Mar 28th 16 - v0.1.4
#18 added JWT - https://github.com/auth0/node-jsonwebtoken
Updated Settings , frontend , backend... basically everything


#### Feb 22nd 16 - v0.1.3
#16 Update the aggregation to not mangle the js & to update the variables to handle the fonts paths
#15 Added the a basic socket.io server with angular factory


#### Jan 25th 16 - v0.1.2
#14 Updated the buildreq with its middleware


#### Jan 23rd 16 - v0.1.1
- #11 - inject mongoose so that the query builder 
- #12 - cut of the redundant question of do you want a custome schema
- #13 - fix the small reference issue to get the blog populating again


#### Jan 11th 16 - v0.1.0
## Features
- Customizable CLI for scaffolding of modules
- Can Build Dynamic Api's based off Schema
- Can Open Dynamic Query with routes
- LiveReload & Recompile in Development Env
- Not Dependent on Grunt or Gulp 
- No Magic Wand
- Uses JS Standard & John Papa Angular Style
- UI framework agnostic - Use Boostrap , Materialize or Foundation
- Supports CSS, SCSS & LESS

configs/settings.js
require('dotenv').config({silent: true})

var path = require('path')
var _ = require('lodash')
var environment = require('./environment.js').get()
var baseLine = {
  app: {
    name: process.env.APP_NAME || 'MeanStackJS'
  },
  cache: {
    maxAge: 0,
    headers: true
  },
  security: {
    cors: {
      active: true,
      options: {
        // origin: function (origin, callback) {
        //   var originIsWhitelisted = ['http://localhost:3000', 'localhost:3000'].indexOf(origin) !== -1
        //   callback(null, originIsWhitelisted)
        // }
      },
      preflight: true
    },
    contentLength: {
      active: true,
      options: {
        max: 99999,
        status: 400,
        message: 'Please make a small payload'
      }
    },
    helmet: {
      active: true,
      options: {}
    },
    hpp: {
      active: true,
      options: {}
    },
    contentSecurityPolicy: {
      active: false,
      options: {}
    },
    hpkp: {
      active: false,
      options: {}
    },
    throttler: {
      active: false,
      options: {}
    }
  },
  minify: 'default',
  render: {
    cli: 'lodash', // __ or ejs or lodash.
    seo: 'lodash', // ejs or lodash. default is lodash
    lodash: {
      options: {} // https://lodash.com/docs#template
    },
    ejs: {
      options: {} // https://www.npmjs.com/package/ejs#options
    }
  },
  env: environment,
  // Root path of server
  root: path.join(__dirname, '/../../..'),
  // Server IP
  ip: process.env.IP || '0.0.0.0',
  hostname: process.env.HOST || process.env.HOSTNAME || 'localhost',
  // Enable Swagger.io at localhost:[port]/api/
  swagger: true,
  // Enable the use of babel for ES6
  babel: {
    options: {
      // NOTE you must install the proper package to use here
      presets: [
        'babel-preset-es2015'
      ],
      plugins: [
        'transform-class-properties'
      ]
    },
    folder: 'dist',
    active: false
  },
  // Plato
  plato: {
    title: 'mean stack',
    eslint: {
      lastsemic: true,
      asi: true
    }
  },
  // JWT Object https://github.com/auth0/node-jsonwebtoken
  jwt: {
    // is used to compute a JWT SIGN
    secret: process.env.JWT_SECRET || 'MEANSTACKJS',
    options: {
      expiresIn: 86400 // 24 hours.   or  the old way//60 * 120 // 60 seconds * 120  = 2 hours
    }
  },
  // is used to compute a session hash
  sessionSecret: process.env.SESSION_SECRET || 'MEANSTACKJS',
  // The name of the MongoDB collection to store sessions in
  sessionCollection: 'sessions',
  // The session cookie settings
  sessionCookie: {
    path: '/',
    httpOnly: true,
    // If secure is set to true then it will cause the cookie to be set
    // only when SSL-enabled (HTTPS) is used, and otherwise it won't
    // set a cookie. 'true' is recommended yet it requires the above
    // mentioned pre-requisite.
    secure: false,
    // Only set the maxAge to null if the cookie shouldn't be expired
    // at all. The cookie will expunge when the browser is closed.
    maxAge: null
  },
  sessionName: 'session.id',
  // Supports MAX CDN
  maxcdn: {
    companyAlias: process.env.MAXCDN_COMPANY_ALIAS || '',
    consumerKey: process.env.MAXCDN_CONSUMER_KEY || '',
    consumerSecret: process.env.MAXCDN_CONSUMER_SECRET || ''
  },
  // SEO - Default html setup
  googleAnalytics: process.env.GOOGLE_ANALYTICS || 'UA-71654331-1',
  html: {
    title: process.env.HTML_META_TITLE || 'Mean Stack JS Demo',
    keywords: process.env.HTML_META_KEYWORDS || 'MEAN, MEANSTACKJS, mongodb, expressjs, angularjs,nodejs, javascript',
    description: process.env.HTML_META_DESCRIPTION || 'Mean Stack JS was built for ease of use with javascript at its core. MeanStackJS is a full stack javascript framework that will give you the power to develop web applications',
    ogUrl: process.env.HTML_META_OG_URL || 'https://meanstackjs.herokuapp.com/',
    ogType: process.env.HTML_META_OG_TYPE || 'website',
    ogTitle: process.env.HTML_META_OG_TITLE || 'Mean Stack JS Demo',
    ogDescription: process.env.HTML_META_OG_DESCRIPTION || 'Mean Stack JS was built for ease of use with javascript at its core. MeanStackJS is a full stack javascript framework that will give you the power to develop web applications',
    ogImage: process.env.HTML_META_OG_IMAGE || 'http://meanstackjs.com/images/logo/header.png',
    fbAppId: process.env.HTML_META_FB_APPID || '1610630462580116',
    twitterCreator: process.env.HTML_META_TWITTER_CREATOR || '@greenpioneerdev',
    twitterCard: process.env.HTML_META_TWITTER_CARD || 'summary_large_image',
    twitterTitle: process.env.HTML_META_TWITTER_TITLE || 'Mean Stack JS Demo',
    twitterDescription: process.env.HTML_META_TWITTER_DESCRIPTION || 'Mean Stack JS was built for ease of use with javascript at its core. MeanStackJS is a full stack javascript framework that will give you the power to develop web applications',
    twitterUrl: process.env.HTML_META_TWITTER_URL || 'https://meanstackjs.herokuapp.com/',
    twitterImage: process.env.HTML_META_TWITTER_IMAGE || 'http://meanstackjs.com/images/logo/header.png',
    twitterSite: process.env.HTML_META_TWITTER_SITE || '@meanstackjs',
    canonical: process.env.HTML_META_CANONICAL || 'https://meanstackjs.herokuapp.com/',
    author: process.env.HTML_META_AUTHOR || 'Green Pioneer Solutions'
  },
  seo: require('./seo.js'),
  // AGGREGATION
  // bower_components -  Needs to be manually added below
  // modules - aggregated automatically
  // images - manually called in files
  // styles - manually called  & automatically compiles the global style scss in COMPILED Folder
  // uploads - Automatic uploads to be manually called in the files
  // USE EXTERNAL FILES - 'https://ajax.googleapis.com/ajax/libs/jquery/1.11.3/jquery.min.js',
  // OR USE INTERNAL FILES - '/bower_components/jquery/dist/jquery.js'
  assets: {
    js: [
      '/angular-jwt/dist/angular-jwt.js',
      '/socket.io-client/dist/socket.io.js',
      '/ng-file-upload/dist/ng-file-upload-all.js',
      '/angular-mocks/angular-mocks.js',
      '/angular-cookies/angular-cookies.js',
      '/angular-sanitize/angular-sanitize.js',
      '/angular-animate/angular-animate.js',
      '/angular-resource/angular-resource.js',
      '/angular-ui-router/release/angular-ui-router.js',
      '/angular-ui-bootstrap/dist/ui-bootstrap-tpls.js',
      '/angular-moment/angular-moment.js',
      '/moment/moment.js',
      '/lodash/lodash.js',
      '/toastr/toastr.js',
      '/bootstrap-sass/assets/javascripts/bootstrap.js',
      '/angular/angular.js',
      '/jquery/dist/jquery.js'
    ],
    css: [
      '/styles/compiled/global.style.css',
      '/toastr/build/toastr.css'
    ],
    copy: [
      '/font-awesome/fonts/fontawesome-webfont.woff2',
      '/font-awesome/fonts/fontawesome-webfont.ttf',
      '/bootstrap-sass/assets/fonts/bootstrap/glyphicons-halflings-regular.woff2',
      '/bootstrap-sass/assets/fonts/bootstrap/glyphicons-halflings-regular.ttf'
    ]
  },
  bodyparser: {
    json: {limit: '100kb'},
    urlencoded: {limit: '100kb', extended: true}
  },
  expresValidator: {
    customValidators: {
      isArray: function (value) {
        // req.assert('param', 'Invalid Param').isArray()
        return _.isObject(value)
      },
      isObject: function (value) {
        // req.assert('param', 'Invalid Param').isObject()
        return _.isObject(value)
      },
      isString: function (value) {
        // req.assert('param', 'Invalid Param').isString()
        return _.isString(value)
      },
      isRegExp: function (value) {
        // req.assert('param', 'Invalid Param').isRegExp()
        return _.isRegExp(value)
      },
      isEmpty: function (value) {
        // req.assert('param', 'Invalid Param').isEmpty()
        return _.isEmpty(value)
      },
      gte: function (param, num) {
        // req.assert('param', 'Invalid Param').gte(5)
        return _.gte(param, num)
      },
      lte: function (param, num) {
        // req.assert('param', 'Invalid Param').lte(5)
        return _.lte(param, num)
      },
      gt: function (param, num) {
        // req.assert('param', 'Invalid Param').gt(5)
        return _.gt(param, num)
      },
      lt: function (param, num) {
        // req.assert('param', 'Invalid Param').lt(5)
        return _.lt(param, num)
      }
    },
    customSanitizers: {
      toArray: function (value) {
        // req.sanitize('postparam').toArray()
        return _.toArray(value)
      },
      toFinite: function (value) {
        // req.sanitize('postparam').toFinite()
        return _.toFinite(value)
      },
      toLength: function (value) {
        // req.sanitize('postparam').toLength()
        return _.toLength(value)
      },
      toPlainObject: function (value) {
        // req.sanitize('postparam').toPlainObject()
        return _.toPlainObject(value)
      },
      toString: function (value) {
        // req.sanitize('postparam').toString()
        return _.toString(value)
      }
    },
    errorFormatter: function (param, message, value) {
      var namespace = param.split('.')
      var root = namespace.shift()
      var formParam = root

      while (namespace.length) {
        formParam += '[' + namespace.shift() + ']'
      }
      return {
        param: formParam,
        message: message,
        value: value
      }
    }
  },
  buildreq: {
    console: true,
    response: {
      method: 'get',
      data: {},
      user: {},
      count: 0,
      hostname: '',
      type: '',
      actions: {
        prev: false,
        next: false,
        reload: false
      },
      delete: ['error']
    },
    query: {
      sort: '',
      limit: 20,
      select: '',
      filter: {},
      populateId: 'user',
      populateItems: '',
      lean: false,
      skip: 0,
      where: '',
      gt: 1,
      lt: 0,
      in: [],
      equal: '',
      errorMessage: 'Unknown Value'
    },
    routing: {
      schema: true,
      url: '/api/v1/',
      build: true
    }
  },
  email: {
    templates: {
      welcome: {
        subject: process.env.EMAIL_WELCOME_SUB || 'Welcome to Mean Stack JS',
        text: function (username) {
          return process.env.EMAIL_WELCOME_TEXT_STRING || 'Hi ' + username + ',\n\n' +
          'Thanks for signing up for Mean Stack JS.\n\n' +
          'If you have any questions about the site, you can reply to this ' +
          'email.\n\n' +
          '— Mean Stack JS'
        }
      },
      reset: {
        subject: process.env.EMAIL_RESET_SUB || 'Reset your password on MEANSTACKJS ',
        text: function (email) {
          return process.env.EMAIL_RESET_TEXT_STRING || 'Hello,\n\n' +
          'This is a confirmation that the password for your account ' + email + ' has just been changed.\n'
        }
      },
      forgot: {
        subject: process.env.EMAIL_FORGOT_SUB || 'Welcome to Mean Stack JS',
        text: function (host, token) {
          return process.env.EMAIL_FORGOT_TEXT_STRING || 'You are receiving this email because you (or someone else) have requested the reset of the password for your account.\n\n' +
          'Please click on the following link, or paste this into your browser to complete the process:\n\n' +
          'http://' + host + '/reset/' + token + '\n\n' +
          'If you did not request this, please ignore this email and your password will remain unchanged.\n'
        }
      }
    },
    from: process.env.EMAIL_FROM || 'MEANSTACKJS@localhost.com',
    error: process.env.EMAIL_ERROR || 'MEANSTACKJS@localhost.com',
    connect: {
      host: process.env.EMAIL_HOST || 'smtp.mandrillapp.com', // Gmail, SMTP
      port: process.env.EMAIL_PORT || '587',
      auth: {
        user: process.env.EMAIL_USER || 'hackathonstarterdemo',
        pass: process.env.EMAIL_PASS || 'E1K950_ydLR4mHw12a0ldA'
      }
    }
  }
}
if (environment === 'test') {
  baseLine = _.assign(baseLine, require('./environments/test.js'))
} else if (environment === 'production') {
  baseLine = _.assign(baseLine, require('./environments/production.js'))
} else if (environment === 'nightwatch') {
  baseLine = _.assign(baseLine, require('./environments/nightwatch.js'))
} else {
  baseLine = _.assign(baseLine, require('./environments/development.js'))
}

exports.get = function (env) {
  return baseLine
}
exports.set = function (identifer, value) {
  baseLine[identifer] = value
  return baseLine
}
